### DISCUSSION TITLE
Possibility to replace quotation marks or brackets

### DISCUSSION THREAD

**Discussion by dflvunoooooo** [espanso/espanso#1915] [Q&A] [answered]
Is it possible to replace brackets or quotation marks? I would like to replace "" with "$|$" or () with ($|$). The only mention in the docs is, that `' " [] {} > | * & ! % # ` @` are special or reserved, whatever that means.

**Comment by smeech** [comment]
From the [docs](https://espanso.org/docs/matches/basics/#multi-line-expansions), "There are a number of characters that are special (or reserved) and cannot be used as the first character of an [unquoted scalar](https://yaml-multiline.info/): ``' " [] {} > | * & ! % # ` @``". In other words, _if_ they need to appear at the beginning of a line, that line has to be quoted. For this reason, the many expansions that use `replace: "{{output}}"` require quote marks around the string. If there is other text preceding the variable, the quote marks aren't actually required, although many of us use them anyway.

Do I take it you'd like to be able to do something like:
```yml
  - trigger: ""
    replace: $|$
```
?
I'm asking, because:
```yml
  - trigger: ()
    replace: (|)
```
is straightforward, so I've probably misunderstood.

**Reply by dflvunoooooo** [reply]
Thank you for your answer! I am very new to espanso, I didn't know "" wasn't necessary. 

Yes almost. I meant to jump in the middle of "" after typing both. But this isn't doing anything
```
  - trigger: ""
    replace: "$|$"
```
This is because it can not start with ", correct?

And the same for (), I would like to end up with the cursor in between the brackets, but my code
```
  - trigger: ()
    replace: ($|$)
```
will produce this `()∫`, with the cursor at the end.

**Comment by smeech** [comment, marked as answer]
Some people prefer to use quote-marks. I feel they make the code less readable (and would rewrite the docs accordingly!), and they mean that "`\`"s and other characters need to be [escaped](https://yaml-multiline.info/), which makes [regex](https://regex101.com/) testing more difficult.

```yml
  - trigger: '""'
    replace: '"$|$"'
```
and
```yml
  - trigger: ()
    replace: ($|$)
```
will do what you need, I think.

**Reply by dflvunoooooo** [reply]
Thank you. With your suggestions, I now get an integral sign at the end of both replacements.
```
()∫
""∫
```

**Reply by smeech** [reply]
That's odd. Is it something in your clipboard that's being injected? There are some clipboard [options](https://espanso.org/docs/configuration/options/) that might ameliorate that. However I noticed it in your previous comment, so it seems to be too consistent for that to be the problem.

Do you get it with other expansions? It's probably worth trying the different [`force_mode:`](https://espanso.org/docs/matches/basics/#injection-mechanism) options in your expansion, and if they make a difference, consider changing your `backend:` globally or in an [app-specific config](https://espanso.org/docs/configuration/app-specific-configurations/).

What's your OS and keyboard language?

Sorry you're encountering these difficulties at the start of your Espanso experience!

**Reply by dflvunoooooo** [reply]
This really is odd. It is working randomly once, then some time not and than again once or twice. Maybe it is related with the five seconds lag of conversion I am experiencing sometimes? It seems to be working more reliably with `force_mode: clipboard`. Thank you for that. Since I am using wayland, I can not use app-specific config. 

I am using arch linux with kde plasma 6 the keyboard language is german with the neo layout. I had to configure that in espanso config to get it working in general.

No worries, thank you for you fast and good support? :)

**Reply by dflvunoooooo** [reply]
Hm, the lag seems to be gone, a few minutes after the last config update. With the `force_mode: clipboard` the integral sign doesn't appear anymore. Now the malfunction seems to be, that every fifth time or so, the replacement of () or "" does delete those characters.

**Reply by smeech** [reply]
It might be worth increasing the `pre_paste_delay:`.

**Reply by dflvunoooooo** [reply]
Thank you. I increased it to 500 ms. It seems to be working more reliable and the deletion is less often. I will try this for today and report back.

**Reply by dflvunoooooo** [reply]
() and "" do not work consistently, everything else seems to be working better.

**Reply by dflvunoooooo** [reply]
I doesn't seem to be working reliably Linux with Wayland for me, whatever I try.
